Moves zipkin-core language level to JRE 6

codefromthecrypt opened the pull request:
This is an api compatible change to allow code sharing between zipkin
server and instrumentation, and without having to publish 2 jar files.

We accidentally broke the ability to make agent-based instrumentation by
compiling our model classes against Java 7 bytecode. Particularly, this
causes a regression in Brave.

To move back to Java 6 bytecode, we had to do a few things:
- replace moshi internal dependency with gson, which is java 6 compliant
- remove okio dependency and copy its Base64 class 
  - incidentally a class it also copied!
- do variable-size buffer writes w/ byte arrays instead of okio
- fix a few nags about diamond ops etc.

See https://github.com/openzipkin/brave/issues/225

codefromthecrypt commented:
cc @ivansenic

eirslett commented:
How about we make a Java 6-compliant fork of the instrumentation library? zipkin-java6? Separate repo? Supporting Java 5 or 6 is an additional burden to contributors; it could be trivial, it could be nontrivial (I'm not sure), but things like okio/moshi only working with Java 7+ is a problem.

Maintaining the fork would be on a best-effort basis, possibly driven by Zipkin users who still use Java 5/6?

Commit:
Moves zipkin-core language level to JRE 6

This is an api compatible change to allow code sharing between zipkin
server and instrumentation, and without having to publish 2 jar files.

We accidentally broke the ability to make agent-based instrumentation by
compiling our model classes against Java 7 bytecode. Particularly, this
causes a regression in Brave.

To move back to Java 6 bytecode, we had to do a few things:
* replace moshi internal dependency with gson, which is java 6 compliant
* remove okio dependency and copy its Base64 class
  * incidentally a class it also copied!
* do variable-size buffer writes w/ OutputBuffer instead of okio
* fix a few nags about diamond ops etc.

codefromthecrypt commented:
@eirslett so found that gson is still actively supporting java 6, so wasn't bad after using that instead of moshi.

codefromthecrypt commented on zipkin/src/main/java/zipkin/Annotation.java:
```diff
@@ -101,11 +100,6 @@ public Annotation build() {
 
   @Override
-  public String toString() {
-    return JsonCodec.ANNOTATION_ADAPTER.toJson(this);
```
removed toStrings on leave nodes while leaving json toString on entrypoint ones (Span and DependencyLink). This reduces the amount of internal class leak

codefromthecrypt commented:
OK.. done toying around. The new implementation uses gson instead of moshi, and a normal buffer instead of okio.

I ran benchmarks before and after:
- Thrift reads were unaffected. Thrift writes (which used to use okio), are slightly faster with small spans and the same with spans with several annotations.
- Json reads (which now use gson instead of moshi) are 20-35% faster, but writes are about half as fast.

While I'm sure I'll nerd out on json writes, in the future, I think this performance is fine. Those concerned with performance should be using thrift anyway, as it is an order of magnitude faster than json before and after this change.
